### PR TITLE
[SandboxIR][NFC] Introduce templated CastInstImpl to simplify subclasses

### PR DESCRIPTION
**Pull request by aeubanks**
The CastInst subclasses all have pretty much the same implementation. Add a helper templated class to help stamp out the subclasses more succinctly.

### CODE REVIEW

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 1]
```diff
@@ -1431,146 +1431,42 @@ class CastInst : public Instruction {
...
-           cast<Instruction>(From)->getOpcode() == Opcode::PtrToInt;
-  }
-#ifndef NDEBUG
-  void dump(raw_ostream &OS) const final;
-  LLVM_DUMP_METHOD void dump() const final;
-#endif // NDEBUG
+class SIToFPInst final : public CastInstImpl<Instruction::Opcode::SIToFP> {};
+class FPToUIInst final : public CastInstImpl<Instruction::Opcode::FPToUI> {};
+class FPToSIInst final : public CastInstImpl<Instruction::Opcode::FPToSI> {};
+class IntToPtrInst final : public CastInstImpl<Instruction::Opcode::IntToPtr> {
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:1462]: clang-format shrug

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 2]
```diff
@@ -1431,146 +1431,42 @@ class CastInst : public Instruction {
...
-
-  static bool classof(const Value *From) {
-    if (auto *I = dyn_cast<Instruction>(From))
-      return I->getOpcode() == Opcode::FPToUI;
-    return false;
+                       Context &Ctx, const Twine &Name = "") {
+    return create(Src, DestTy, InsertBefore->getIterator(),
+                  InsertBefore->getParent(), Ctx, Name);
   }
-#ifndef NDEBUG
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:1470]: I would keep these for now until we find a better fix for all of them. The issue is that gdb won't work with `I->dump()` if `I` is a pointer to a sub-class of CastInst. You can get it to work by casting it to `sandboxir::CastInst *`, like `(gdb) ((sandboxir::CastInst *)I)->dump()` which is annoying.

**Review by vporpo — APPROVED** [review]
We may need a `Value::dump()` with a big switch case. I will work on that once I add the dump tests.
Anyway, feel free to drop the dumps, I think we can live without them for now.

File: llvm/include/llvm/SandboxIR/SandboxIR.h (hunk 2 above)
Comment (aeubanks) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:1470]: it looks like all `dump()` methods are the same right now, with multi-instructions being different in the future. there's gotta be a better way of deduplicating some of the logic, seeing so many copies of

```
void FooInst::dump(raw_ostream &OS) const {
  dumpCommonPrefix(OS);
  dumpCommonSuffix(OS);
}
```

hurts.

can you test if `B().f()` in
```
struct A {
    void f();
};

struct B : A {
    using A::f;
};
```
works with gdb?

Comment (vporpo) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:1470]: Yeah there is a ton of duplication. What might work is a non-virtual `Value::dump()` that calls virtual `dumpImpl()` that may get overriden by the sub-classes if needed.

I added a `using CastInst::dump` in `SIToFPInst`, but gdb still complains: `Couldn't find method llvm::sandboxir::SIToFPInst::dump`.

**Pull request merged**